File: python/140-WordBreakII.py
"""
Given a non-empty string s and a dictionary wordDict containing a list of non-empty words, add spaces in s to construct a sentence where each word is a valid dictionary word. You may assume the dictionary does not contain duplicate words.

Return all such possible sentences.

For example, given
s = "catsanddog",
dict = ["cat", "cats", "and", "sand", "dog"].

A solution is ["cats and dog", "cat sand dog"].

UPDATE (2017/1/4):
The wordDict parameter had been changed to a list of strings (instead of a set of strings). Please reload the code definition to get the latest changes.
"""

_end = '_end_'

def make_trie(words):
    root = dict()
    for word in words:
        current_dict = root
        for letter in reversed(word):
            current_dict = current_dict.setdefault(letter, {})
        current_dict[_end] = _end
    return root


class Solution(object):
    def wordBreakJudge(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        >>> Solution().wordBreak('leetcode', wordDict=['leet', 'code'])
        True
        >>> Solution().wordBreak('leetcode', wordDict=['leet', 'cod'])
        False
        >>> Solution().wordBreak('leetcode', wordDict=['leet', 'cod', 'e'])
        True
        """
        # first we make a trie data structure for fast search
        reversed_trie = make_trie(wordDict)
        vec = [True] + [False] * len(s)
        for i in range(len(s)):
            words = reversed_trie
            currentIndex = i
            while currentIndex >= 0:
                if s[currentIndex] not in words:
                    break
                else:
                    words = words[s[currentIndex]]
                    if _end in words and vec[currentIndex] == True:
                        vec[i + 1] = True
                        break
                    currentIndex -= 1
        return vec[len(s)], reversed_trie
    
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: List[str]
        >>> s = "catsanddog"
        >>> dictionary = ["cat", "cats", "and", "sand", "dog"]
        >>> set(Solution().wordBreak(s, dictionary)) == set(["cats and dog", "cat sand dog"])
        True
        """
        # sanity check, for case where a char in s is not in dict. This is really boring
        canBeBreak, reversed_trie = self.wordBreakJudge(s, wordDict)
        if canBeBreak is False:
            return []

        # first we make a trie data structure for fast search
        vec_index = [[[-1]]] + [[] for i in range(len(s))]
        for i in range(len(s)):
            words = reversed_trie
            currentIndex = i
            while currentIndex >= 0:
                if s[currentIndex] not in words:
                    break
                else:
                    words = words[s[currentIndex]]
                    if _end in words and len(vec_index[currentIndex]) != 0:
                        for w in vec_index[currentIndex]:
                            vec_index[i + 1].append(w[:] + [i + 1])
                    currentIndex -= 1
        result = []
        for index_lst in vec_index[len(s)]:
            string = s[0:index_lst[1]]
            for i in range(1, len(index_lst) - 1):
                string += ' ' + s[index_lst[i]:index_lst[i+1]]
            result.append(string)
        return result

File: cpp/140.word-break-ii.cpp
/*
 * @lc app=leetcode id=140 lang=cpp
 *
 * [140] Word Break II
 *
 * https://leetcode.com/problems/word-break-ii/description/
 *
 * algorithms
 * Hard (36.21%)
 * Total Accepted:    380.5K
 * Total Submissions: 971.7K
 * Testcase Example:  '"catsanddog"\n["cat","cats","and","sand","dog"]'
 *
 * Given a string s and a dictionary of strings wordDict, add spaces in s to
 * construct a sentence where each word is a valid dictionary word. Return all
 * such possible sentences in any order.
 * 
 * Note that the same word in the dictionary may be reused multiple times in
 * the segmentation.
 * 
 * 
 * Example 1:
 * 
 * 
 * Input: s = "catsanddog", wordDict = ["cat","cats","and","sand","dog"]
 * Output: ["cats and dog","cat sand dog"]
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: s = "pineapplepenapple", wordDict =
 * ["apple","pen","applepen","pine","pineapple"]
 * Output: ["pine apple pen apple","pineapple pen apple","pine applepen apple"]
 * Explanation: Note that you are allowed to reuse a dictionary word.
 * 
 * 
 * Example 3:
 * 
 * 
 * Input: s = "catsandog", wordDict = ["cats","dog","sand","and","cat"]
 * Output: []
 * 
 * 
 * 
 * Constraints:
 * 
 * 
 * 1 <= s.length <= 20
 * 1 <= wordDict.length <= 1000
 * 1 <= wordDict[i].length <= 10
 * s and wordDict[i] consist of only lowercase English letters.
 * All the strings of wordDict are unique.
 * 
 * 
 */

#include "common.hpp"

class Solution {
public:
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        // trie and backtracking
        root = new TrieNode();
        for (auto &word: wordDict) {
            this->add(word);
        }
        std::vector<std::string> result;
        std::string curr;
        this->dfs(result, curr, s, 0, 0, root);
        return result;
    }

private:
    struct TrieNode {
        bool anchor;
        TrieNode **next;
        TrieNode(): anchor(false) {
            next = new TrieNode*[26];
            for (int i = 0; i < 26; i++) {
                next[i] = nullptr;
            }
        }
    };

    TrieNode *root;

    void add(const std::string &word) {
        TrieNode *curr = root;
        for (auto & c: word) {
            int index = c - 'a';
            // std::cout << index << std::endl;
            if (curr->next[index] == nullptr) {
                curr->next[index] = new TrieNode();
            }
            curr = curr->next[index];
        }
        curr->anchor = true;
    }

    void dfs(std::vector<string>&result, std::string &curr, const std::string &s, int start_index, int curr_index, TrieNode* node) {
        // base case
        if (curr_index == s.size()) {
            if (start_index == s.size()) {
                result.push_back(curr);
                return;
            } else {
                return;
            }
        }
        // find next node according to s
        char c = s.at(curr_index);
        TrieNode *next = node->next[c - 'a'];
        if (next == nullptr) return;  // no such word
        if (next->anchor) {
            // split here [start_index, curr_index]
            auto sub_string = s.substr(start_index, curr_index - start_index + 1);

            // std::cout << sub_string << std::endl;

            int inserted_size = 0;
            if (curr.size() != 0) {
                curr += " ";
                inserted_size += 1;
            }
            curr += sub_string;
            inserted_size += sub_string.size();
            this->dfs(result, curr, s, curr_index + 1, curr_index + 1, root);

            // pop back
            for (int i = 0; i < inserted_size; i++) {
                curr.pop_back();
            }

        }
        // not split or continue searching
        this->dfs(result, curr, s, start_index, curr_index + 1, next);
    }
};


int main(int argc, char const *argv[]) {
    std::vector<string> wordList {"cat","cats","and","sand","dog"};
    Solution().wordBreak("catsanddog", wordList);
    return 0;
}
